enhancements/cluster-logging/elasticsearch-v2-api-and-infrastructure.md

```diff
@@ -0,0 +1,310 @@
+in optimal ways and usually requires an over allocation of resources (which makes it hard to justify horizontal scaling, which leads to performance
+related issues). It would also allow us to move away from the idea of node UUIDs and make naming conventions more generic and simple.
+
+### Goals
+The specific goals of this proposal are:
+
+* Outline the responsibility for the Elasticsearch Operator regarding desired infrastructure
+
+* Determine a v2 elasticsearch CRD API
+
+* Discuss upgrade strategies from v1 to v2
+
+We will be successful when:
+
+* Users of Elasticsearch Operator no longer need to specify their node roles and count
+
+### Non-Goals
+
+* N/A
+
+### Current workflow vs Proposed
+
+Currently, a consumer like Jaeger or Cluster Logging is responsible for the following for their ES cluster:
+
+1. Creating and updating their elasticsearch CR(s) so that EO can manage their cluster
+    * This includes breaking up nodes by roles
+    * Specifying (usually pass through) of node requirements
+
+With this proposal the workflow would be:
+
+1. Create their elasticsearch CR specifying a particular sized cluster (based on t-shirt sizes)
+
+Current specification:
+
+```yaml
+apiVersion: "logging.openshift.io/v1"
+kind: "Elasticsearch"
+metadata:
+  name: "elasticsearch"
+spec:
+  managementState: "Managed"
+  nodeSpec:
+    resources:
+      limits:
+        memory: 1Gi
+      requests:
+        cpu: 100m
+        memory: 1Gi
+  nodes:
+  - nodeCount: 1
+    roles:
+    - client
+    - data
+    - master
+    storage:
+      storageClassName: gp2
+      size: 10Gi
+  redundancyPolicy: ZeroRedundancy
+  indexManagement:
+    policies:
+    - name: infra-policy
+      pollInterval: 1m
+      phases:
+        hot:
+          actions:
+            rollover:
+              maxAge:   2m
+        delete:
+          minAge: 5m
+    mappings:
+    - name:  infra
+      policyRef: infra-policy
+      aliases:
+      - infra
+      - logs.infra
+```
+
+Proposed:
+
+```yaml
+apiVersion: "logging.openshift.io/v1"
+kind: "Elasticsearch"
+metadata:
+  name: "elasticsearch"
+spec:
+  storage:
+    storageClassName: gp2
+    size: 10Gi
+  size: 0x.medium
+  indexManagement:
+    infraPolicies:
+    - phases:
+      hot:
+        actions:
+          rollover:
+            maxAge:   2m
+      delete:
+        minAge: 5m
+    appPolicies:
+    - phases:
+      hot:
+        actions:
+          rollover:
+            maxAge: 2m
+      delete:
+        minAge: 5m
+```
+
+## Proposal
+
+Currently the Elasticsearch CR is overly complicated and allows for too much granular control over the data nodes, which places the expectation
+of expertise on the user rather than containing it to the operator (which is where it should reside). This proposal seeks to outline a v2 to provide
+a more simple means of interracting with the Elasticsearch Operator to ask for a managed Elasticsearch cluster that can be better
+configured for performance.
```

Comment:
> Currently the Elasticsearch CR is overly complicated and allows for too much granular control over the data nodes, which places the expectation of expertise on the user rather than containing it to the operator (which is where it should reside).

This feels like a motivation statement.  Not sure this is really proposing anything.

Perhaps a proposal statement could be: "Provide a v2 API which simplifies the configuration YAML to provide storage size, data retention times, and a "t-shirt" size summary configuration.

> This proposal seeks to outline a v2 to provide a more simple means of interacting with the Elasticsearch Operator to ask for a managed Elasticsearch cluster that can be better configured for performance.

It is not clear to me how t-shirt sizes can work when you allow the user to specify storage size and data retention times.  Seems like that still allows them to shoot themselves in the foot.

Reply:
If we know our data retention goals the operator can ensure it is scaled up enough based on the resources it is allowed to allocate per node. that, or it tells the users that there is no way it can meet that goal... which would be in the form of an alert